### DISCUSSION TITLE
Reading files with uproot and xrootd: how to set timeout?

### DISCUSSION THREAD

**Discussion by andreypz** [scikit-hep/coffea#585] [Q&A] [answered]
Hello,

I would like to use `processor.run_uproot_job()` method and open multiple files located on xrootd. It works for a few files, but crashes when I try more.
I think it's related to the fact the the xrd.timeout is not set (I can see that by printing out the value in uptoot/source/xrootd.py - it returns `None`). So my question is how can I set it in `processor.run_uproot_job()`?

**Comment by lgray** [comment, marked as answer]
@andreypz It would be helpful to see how things are crashing in your case. It could be due to a variety of things beyond the timeout. Typically the lack of a timeout causes stalls and not crashes. In particular you should try telling `run_uproot_job` to skip bad files: https://github.com/lgray/hgg-coffea/blob/master/runner.py#L355.

To make a timeout really sticky you need to set it when the dask/parsl/etc worker job is started. If you are running with Dask you should add the timeouts via dask to the worker environment. You need to set `XRD_REQUESTTIMEOUT` and `XRD_STREAMTIMEOUT` to some reasonable values. For the local executors you can set the environment variables yourself or use the `xrootdtimeout` executor config option.

```python3
run_uproot_job(..., executor_args={"xrootdtimeout": Nseconds, ...}, ...)
```

**Reply by andreypz** [reply]
Thanks for the answer. I run with local executor and the `"xrootdtimeout"` setting  is indeed propagated to xrootd. However this does not solve the crash. So it's not the timeout indeed.

The crash I get is this one:
```
FileNotFoundError: file not found ([ERROR] Server responded with an error: [3006] tried hosts option not supported.)

    'root://xrootd-cms.infn.it///store/mc/RunIIFall17NanoAODv7/DY2JetsToLL_M-50_LHEZpT_50-150_TuneCP5_13TeV-amcnloFXFX-pythia8/NANOAODSIM/PU2017_12Apr2018_Nano02Apr2020_102X_mc2017_realistic_v8-v1/70000/9C48232C-28E4-7D4F-B5C4-36A06B1906FD.root'
```
Any ideas what could be the problem?
This file can be opened with `root` without any problem.

**Reply by andreypz** [reply]
Setting `"skipbadfiles": True` does not work for me. It crashes with:
```
    output = processor.run_uproot_job(file_list,
  File "/.automount/home/home__home1/institut_3a/andreypz/work/vjets/gen-scripts/geneno/lib/python3.8/site-packages/coffea/processor/executor.py", line 1222, in run_uproot_job
    processor_instance.postprocess(wrapped_out["out"])
TypeError: 'NoneType' object is not subscriptable
```

**Reply by lgray** [reply]
It's the extra "/" in `root://xrootd-cms.infn.it///store/mc/...` where properly formatted it should be `root://xrootd-cms.infn.it//store/mc/...`. I suspect root or something else is taking liberties with what the LFN string means and what can you can do to normalize it.

If I make that edit the file opens with no problem in uproot from LPC, using infn redirector.

@jpivarski - is this expected?

**Reply by lgray** [reply]
> Setting `"skipbadfiles": True` does not work for me. It crashes with:
> 
> ```
>     output = processor.run_uproot_job(file_list,
>   File "/.automount/home/home__home1/institut_3a/andreypz/work/vjets/gen-scripts/geneno/lib/python3.8/site-packages/coffea/processor/executor.py", line 1222, in run_uproot_job
>     processor_instance.postprocess(wrapped_out["out"])
> TypeError: 'NoneType' object is not subscriptable
> ```

@nsmith- bug?

**Reply by nsmith-** [reply]
I think that happens when _all_ files are skipped, hence the output is null. Could be handled better

**Reply by andrzejnovak** [reply]
Yup that's a common question that comes up

**Reply by jpivarski** [reply]
I don't know how many slashes are needed in an XRootD URL, but Uproot does not modify that.

One thing to know about XRootD timeouts is that they must be integers, and hence whole numbers of seconds. If you're trying to set a fraction of a second, it probably gets cast as an int and therefore becomes 0, which I think means "no timeout."

**Reply by andreypz** [reply]
Thanks @lgray , changing to two slashes  seems to fix this one. Strange that three slashes worked fine for all other files I run over. Anyway, turns out timeout is irrelevant for this issue.
